pip/configuration.py

```diff
@@ -4,68 +4,245 @@
 import re
 import os
 import sys
+import logging
 
+from pip._vendor.six import next
 from pip._vendor.six.moves import configparser
+
 from pip.locations import (
-    legacy_config_file, config_basename, running_under_virtualenv,
-    site_config_files
+    new_config_file, legacy_config_file, venv_config_file,
+    running_under_virtualenv, site_config_files
 )
-from pip.utils import appdirs
+from pip.utils import ensure_dir
 
 
 _environ_prefix_re = re.compile(r"^PIP_", re.I)
+_need_file_err_msg = "Needed a specific file to be modifying."
+
+
+logger = logging.getLogger(__name__)
+
+
+# NOTE: Maybe use the optionx attribtue to normalize keynames.
+def _normalize_name(name):
+    """Make a name consistent regardless of source (environment or file)
+    """
+    name = name.lower().replace('_', '-')
+    if name.startswith('--'):
+        name = name[2:]  # only prefer long opts
+    return name
+
+
+def _disassemble_key(name):
+    return name.split(".", 1)
+
+
+def _make_key(variant, name):
+    return ".".join((variant, name))
+
+# Some terminology:
```

Comment:
Should this stay separate as a comment or be added to a doc-string?

Maybe, because it involves terminology only for this module - keep it as it is.